design/authenticators/resource-restrictions/resource-restrictions.md

```diff
@@ -0,0 +1,350 @@
+
+#### 2. Validate the Resource Restrictions are defined correctly
+
+In the Azure Authenticator and in the K8s Authenticator we have constraints on 
+how the annotations are defined.
+
+ As can be seen 
+[here](/app/domain/authentication/authn_azure/validate_resource_restrictions.rb), 
+we have the following validations:
+
+  - Validate only permitted restrictions are defined
+    - Here we validate that the `authn-azure` annotations are known. For
+      example, the annotation `authn-azure/subscription-id` is valid, whereas
+      the annotation `authn-azure/unknown` is invalid. We do that to minimize
+      errors from the user, where they will define, for example, the
+      annotation `authn-azure/subscription-ir` (note the typo). In this
+      case we'd rather fail the authentication and point them to their
+      mistake rather than ignore it, and authenticate the request in a weaker
+       granularity than what is expected by the user.
+  - Validate required restrictions exist
+    - For example, in the Azure Authenticator we require hosts to have the
+     `subscription-id` and the `resource-group` annotations. In the K8s
+      Authenticator we require the `namespace` annotation.
+  - Validate mutually exclusive restrictions
+    - For example, in the Azure Authenticator we allow hosts to have
+     only one of the `user-assigned-identity` and the `system-assigned-identity`
+     annotations. In the K8s we allow hosts to have only one of
+      the annotations `deployment`, `deployment-config` & `stateful-set`.
+
+Also here we can see some generalization.
+
+These need to be defined for each authenticator.
+To simplify that, we can define a simple class for each of these constraints.
+The whole constraints will be an array of these constraints, defined once for each authenticator.
+Then, to validate the Resource Restrictions, it only needs to loop on these constraints, and invoke
+`constraint.validate(resource_restrictions)` on each.
+
+This way, adding or changing the logic is more simple and intuitive, while maintaining readability.
+
+Each of the constraints classes will have the same `validate` signature, and will look similar to:
+
+```ruby
+class RequiredConstraint
+
+  def initialize(required:)
+    @required = required
+  end
+
+  def validate(resource_restrictions:)
+    # implement the constraint
+  end
+
+end
+```
+
+It will be used by each authenticator to define how to validate the configured
+restrictions. These are the ones that were defined in the policy for the host,
+regardless of the request.
+
+Example how it will look like for `authn-azure`:
+
+```ruby
+REQUIRED = [SUBSCRIPTION_ID, RESOURCE_GROUP]
+IDENTITY_EXCLUSIVE = [USER_ASSIGNED_IDENTITY, SYSTEM_ASSIGNED_IDENTITY]
+PERMITTED = REQUIRED + IDENTITY_EXCLUSIVE
+
+CONSTRAINTS =
+    [
+        RequiredConstraint.new(required: REQUIRED),
+        PermittedConstraint.new(permitted: PERMITTED),
+        ExclusiveConstraint.new(exclusive: IDENTITY_EXCLUSIVE)
+    ]
```

Comment:
A small point, but I don't see much benefit to introducing indirection via constants here.  That is, I'd rather see an object with actual values and no indirection, something more or less like (I'm handwaving over details/names):
```ruby
CONSTRAINTS = Constraints.new(
  required: [:subscription_id, :resource_group],
  permitted: [:user_assigned_identity, system_assigned_identity],
  exclusive: [:foo, :bar]
)
```
In case I'm missing something, the way I'm thinking about this is asking, "What is the reason to use constants like these?"  There are two typical answers:
1. We need to expose the values to other code.
2. We want a layer of indirection so that we can refer to the constant in multiple places within our class, but update its value in a single place.  That is, if the value of the constant changes, we want to change it once, and have that reflected everywhere.

However, if you only use the constant's value in one place, reason 2 does not apply, and it's better not to use the constant in that case, because the indirection has a cost in readability and simplicity.  You only want to pay that cost when you have to.

Reply:
> You only want to pay that cost when you have to.

I totally agree.
But when I looked deeply into the code I saw that actually, both of the reasons apply here.
1. These constants represent strings and they are used to create a mapping to values in the request.
That's the 3rd step, where we compare the resource restrictions values with the request values.
2. `Permitted` is the sum of all other constraints, plus optional if there are any.
So defining them once prevents the necessity to write the same strings multiple times.